src/ipc/handler.c

```diff
@@ -449,6 +449,18 @@ static int ipc_stream_trigger(uint32_t header)
 		trace_error_value(ipc_cmd);
 	}
 
+	/* when stop command comes, the pipeline has to be prepared.
+	 * This case has to be taken into account:
+	 * when ALSA has XRUN, it will send stop/start commands to FW
+	 * and try to recover the pipepine. If we did not prepare
+	 * the pipeline before start, the recover process will fail.
+	 */
+	if (cmd == COMP_TRIGGER_STOP) {
+		ret = pipeline_prepare(pcm_dev->cd->pipeline, pcm_dev->cd);
```

Comment:
that looks quite questionable. it's rather when you start that you should make sure things are prepared, no?

Reply:
It takes effect similar like the pipeline XRUN recover process.
you can refer to the pipeline_xrun_recover() to understand the logic.
without prepare, the firmware will hit XRUN again when start command comes.

Reply:
I make a stupid typo error, the Line459 should be:
ret = pipeline_prepare(pcm_dev->pipeline, pcm_dev->pipeline->source_comp);
I will update it after back office.